app/src/main/java/com/fstyle/structure_android/utils/Validator.java

```diff
@@ -0,0 +1,277 @@
+                Object real = field.get(model);
+                // detect when data has changed only if onlyValidateTheChange is true.
+                // Otherwise, always run validation on all of fields .
+                if (!onlyValidateChange || (cache == null && real != null) || (cache != null
+                        && cache.equals(real))) {
+                    validate(real, rules, isOptional);
+                }
+            } catch (IllegalAccessException e) {
+                Log.e(TAG, "validate: ", e);
+            }
+            mModelCache = model.clone();
+        }
+    }
+
+    @StringRes
+    public int getErrorByTypeInt(@ValidType int type) {
+        int index = mErrorTraces.indexOfValue(type);
+        if (index == -1) {
+            return R.string.empty;
+        }
+        return mErrorTraces.keyAt(index);
+    }
+
+    public String getErrorByType(@ValidType int type) {
+        int index = mErrorTraces.indexOfValue(type);
+        if (index == -1 || mErrorTraces.keyAt(index) == -1) {
+            return null;
+        }
+        return mContext.getString(mErrorTraces.keyAt(index));
+    }
+
+    public void initNGWordPattern() {
+        Observable.create(new Observable.OnSubscribe<Pattern>() {
+            @Override
+            public void call(Subscriber<? super Pattern> subscriber) {
+                try {
+                    BufferedReader reader = new BufferedReader(
+                            new InputStreamReader(mContext.getAssets().open("ng-word")));
+                    StringBuilder buffer = new StringBuilder();
+                    String line;
+                    while ((line = reader.readLine()) != null) {
+                        if (buffer.length() == 0) {
+                            buffer.append("(").append(line.toLowerCase());
+                        } else {
+                            buffer.append("|").append(line.toLowerCase());
+                        }
+                    }
+                    subscriber.onNext(Pattern.compile(buffer.append(")").toString()));
+                } catch (IOException e) {
+                    subscriber.onError(e);
+                }
+            }
+        })
+                .subscribeOn(Schedulers.io())
+                .observeOn(AndroidSchedulers.mainThread())
+                .subscribe(new Action1<Pattern>() {
+                    @Override
+                    public void call(Pattern pattern) {
+                        mNGWordPattern = pattern;
+                    }
+                });
+    }
+
+    @ValidMethod(type = { ValidType.NG_WORD })
+    public boolean validateNGWord(String str) {
+        if (mNGWordPattern == null) {
+            throw new ValidationException(
+                    "NGWordPattern is null!!! \n Call initNGWordPattern() before call this method!",
+                    new NullPointerException());
+        }
+        return TextUtils.isEmpty(str) || !mNGWordPattern.matcher(str.toLowerCase()).find();
```

Comment:
WARNING: [DefaultLocale] Implicitly using the default locale is a common source of bugs: Use `toLowerCase(Locale)` instead
 Calling `String#toLowerCase()` or `#toUpperCase()` *without specifying an explicit locale* is a common source of bugs. The reason for that is that those methods will use the current locale on the user's device, and even though the code appears to work correctly when you are developing the app, it will fail in some locales. For example, in the Turkish locale, the uppercase replacement for `i` is *not* `I`.

If you want the methods to just perform ASCII replacement, for example to convert an enum name, call `String#toUpperCase(Locale.US)` instead. If you really want to use the current locale, call `String#toUpperCase(Locale.getDefault())` instead.